python/cudf/cudf/_lib/scalar.pyx

```diff
@@ -43,7 +43,6 @@ from cudf._lib.cpp.scalar.scalar cimport (
 )
 cimport cudf._lib.cpp.types as libcudf_types
 
-
 cdef class Scalar:
```

Comment:
The caching logic looks like it can live in Python-land -- the underlying `_host_object` and `_host_dtype` attributes are plain Python object's and don't really benefit from Cython typing. Any particular reason we keep that logic in Cython? Python code is easier to debug and making Python changes doesn't need the Cython to be rebuilt

Reply:
I think the main issue is basically that the code still uses `cudf._lib.scalar.Scalar` in a few places. So until those instances are resolved, the cdef class has to have all the attributes/methods that it currently has. But let me double check on that.

Reply:
If that's the case, we should just be able to change all the imports to just `cudf.core.scalar.Scalar`. Everything _should_ still just work, since Python `Scalar` inherits Cython `Scalar`, right?

Reply:
I suspect so - will give that a shot.

Reply:
I don't think this involves large changes, but in case it does, let's `#TODO` it for a follow-up. But I do strongly suggest we limit Cython to just the part that needs to interact with C++ objects.

Reply:
@shwina just an update on this. I think this might not be quite as simple as I had originally thought. I was able to move things around so that instead of an inheritance relationship between `Scalar` and `DeviceScalar`, we store a `DeviceScalar` as `_data` and forward various methods to the `DeviceScalar`'s methods. I have not been able to make a ton of headway with moving the caching mechanism to the python layer. 

The roadblock comes down to the fact that in the end, libcudf needs to access the c++ scalar object through a cdef class. This means it needs to receive that object from a method of that cdef class (`get_raw_ptr`). Since `get_raw_ptr` needs to set the actual device value from the host value at some point during execution, it needs to be able to see the `_host_value` object. Unless we had a `_parent` or `_owner` attribute, it has no way of seeing that value.